Invalidate cache after a non-safe method on the same URI

adrian3d opened the pull request:
Related to #22 

Deletion of cache data from the request object has been added in the non-safe HTTP method condition.

Kevinrob commented:
Thank you very much for opening this PR! I will review it now.

Kevinrob commented on src/CacheMiddleware.php:
```diff
@@ -113,6 +113,10 @@ public function __invoke(callable $handler)
                 // No caching for this method allowed
+
+                // Invalidate cache after a non-safe method on the same URI will be invalidated
+                $this->cacheStorage->delete($request);
```
> A cache MUST invalidate the effective Request URI (Section 5.5 of
> [RFC7230]) as well as the URI(s) in the Location and Content-Location
> response header fields (if present) when a non-error status code is
> received in response to an unsafe request method.
> [...]
> Here, a "non-error response" is one with a 2xx (Successful) or 3xx 
> (Redirection) status code.
https://tools.ietf.org/html/rfc7234#section-4.4

I thing that it's better if this action is done after checking that the response are "non-error". Can we move it in the `then(...)` just below?

adrian3d replied:
> > A cache MUST invalidate the effective Request URI (Section 5.5 of
> > [RFC7230]) as well as the URI(s) in the Location and Content-Location
> > response header fields (if present) when a non-error status code is
> > received in response to an unsafe request method.
> > [...]
> > Here, a "non-error response" is one with a 2xx (Successful) or 3xx
> > (Redirection) status code.
> > https://tools.ietf.org/html/rfc7234#section-4.4
> 
> I thing that it's better if this action is done after checking that the response are "non-error". Can we move it in the `then(...)` just below?

Makes sense.

adrian3d commented:
> I just added a little comment.
> Can you add a small test about that too?

I've added tests than introduce some changes to improve readability. Hope this will be satisfying.

Kevinrob commented:
Much better! Thank you for your contribution 💪
